File: server/app/ark.py
"""
ARK (火山引擎方舟) 客户端：SSE 透传

前端把 OpenAI 兼容的 chat/completions 请求体发到 /api/ai/chat，
后端补上 Authorization 头，再把上游 SSE 字节流原样转发回浏览器。
"""
from __future__ import annotations

import json
import logging
from typing import Any, AsyncIterator

import httpx

from .config import get_settings

log = logging.getLogger("ark")


async def stream_chat(payload: dict[str, Any]) -> AsyncIterator[bytes]:
    """以 SSE 透传方式调用 ARK chat/completions。

    payload 必须包含 messages，stream 会被强制为 True。
    model 缺省时使用 ARK_DEFAULT_MODEL。
    """
    settings = get_settings()
    if not settings.configured:
        raise RuntimeError("ARK_API_KEY 未配置，请在 server/.env 填写")

    body = dict(payload)
    body["stream"] = True
    body.setdefault("model", settings.ark_default_model)

    url = f"{settings.ark_base_url}/chat/completions"
    headers = {
        "Authorization": f"Bearer {settings.ark_api_key}",
        "Content-Type": "application/json",
        "Accept": "text/event-stream",
    }

    log.info("ARK request → model=%s url=%s", body.get("model"), url)

    timeout = httpx.Timeout(settings.request_timeout_seconds, connect=10.0)
    async with httpx.AsyncClient(timeout=timeout) as client:
        async with client.stream("POST", url, json=body, headers=headers) as resp:
            if resp.status_code >= 400:
                text = (await resp.aread()).decode("utf-8", errors="replace")
                log.error("ARK upstream error %s: %s", resp.status_code, text)
                err = json.dumps(
                    {"status": resp.status_code, "message": text[:500]},
                    ensure_ascii=False,
                )
                yield f"data: {{\"error\": {err}}}\n\n".encode("utf-8")
                yield b"data: [DONE]\n\n"
                return
            async for chunk in resp.aiter_bytes():
                if chunk:
                    yield chunk

File: src/services/ark.ts
/**
 * AI service —— 豆包 / Volc Engine ARK
 * 协议：OpenAI 兼容（/chat/completions），SSE 流式输出
 *
 * 文档：https://www.volcengine.com/docs/82379/1298454
 *
 * 双模式：
 *   - 后端模式（推荐）：VITE_ARK_BASE_URL = /api/ai，由 FastAPI 持有 Key 透传
 *   - 直连模式（仅 dev）：VITE_ARK_BASE_URL = /ark，浏览器直接持 Key 经 vite proxy
 */

export interface ChatMessage {
  role: 'system' | 'user' | 'assistant'
  content: string
}

export interface ArkConfig {
  /** 后端模式下可为空 */
  apiKey: string
  baseUrl: string
  model: string
  /** true 表示走自家后端（/api/ai 前缀），不需要前端持 Key */
  useBackend: boolean
}

export class ArkConfigError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ArkConfigError'
  }
}

function isBackendUrl(baseUrl: string): boolean {
  return /^\/api(\/|$)/.test(baseUrl)
}

export function getArkConfig(): ArkConfig {
  const apiKey = import.meta.env.VITE_ARK_API_KEY ?? ''
  const baseUrl = import.meta.env.VITE_ARK_BASE_URL || '/api/ai'
  const model = import.meta.env.VITE_ARK_MODEL || 'doubao-1-5-pro-32k-250115'
  const useBackend = isBackendUrl(baseUrl)
  if (!useBackend && !apiKey) {
    throw new ArkConfigError(
      '直连模式缺少 VITE_ARK_API_KEY。建议改用后端：将 VITE_ARK_BASE_URL 设为 /api/ai 并启动 server/。'
    )
  }
  return { apiKey, baseUrl, model, useBackend }
}

export function isArkConfigured(): boolean {
  const baseUrl = import.meta.env.VITE_ARK_BASE_URL || '/api/ai'
  if (isBackendUrl(baseUrl)) return true // 后端是否真的就绪由请求阶段验证
  return Boolean(import.meta.env.VITE_ARK_API_KEY)
}

interface StreamOptions {
  messages: ChatMessage[]
  /** 每次模型增量 token 触发 */
  onDelta?: (delta: string, full: string) => void
  /** 完成时触发 */
  onDone?: (full: string) => void
  /** 错误 */
  onError?: (err: Error) => void
  signal?: AbortSignal
  temperature?: number
  /** 自定义额外 messages 之上的覆盖项 */
  extra?: Record<string, unknown>
}

/**
 * 流式调用豆包 chat completions
 * 返回 promise，resolve 时附带完整文本
 */
export async function streamChat(opts: StreamOptions): Promise<string> {
  const cfg = getArkConfig()
  const url = `${cfg.baseUrl}/chat/completions`

  const headers: Record<string, string> = {
    'Content-Type': 'application/json'
  }
  if (!cfg.useBackend) {
    // 直连模式才在浏览器侧带 Key
    headers.Authorization = `Bearer ${cfg.apiKey}`
  }

  const res = await fetch(url, {
    method: 'POST',
    headers,
    signal: opts.signal,
    body: JSON.stringify({
      model: cfg.model,
      messages: opts.messages,
      stream: true,
      temperature: opts.temperature ?? 0.7,
      ...(opts.extra ?? {})
    })
  })

  if (!res.ok || !res.body) {
    const text = await res.text().catch(() => '')
    const err = new Error(`Ark API ${res.status}: ${text || res.statusText}`)
    opts.onError?.(err)
    throw err
  }

  const reader = res.body.getReader()
  const decoder = new TextDecoder('utf-8')
  let buffer = ''
  let full = ''

  try {
    while (true) {
      const { value, done } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })

      // SSE：逐行解析，每条事件以空行分隔
      const lines = buffer.split('\n')
      buffer = lines.pop() ?? ''

      for (const raw of lines) {
        const line = raw.trim()
        if (!line || !line.startsWith('data:')) continue
        const payload = line.slice(5).trim()
        if (payload === '[DONE]') {
          opts.onDone?.(full)
          return full
        }
        try {
          const json = JSON.parse(payload)
          if (json?.error) {
            const msg =
              typeof json.error === 'string'
                ? json.error
                : json.error?.message || JSON.stringify(json.error)
            const err = new Error(`ARK upstream: ${msg}`)
            opts.onError?.(err)
            throw err
          }
          const delta: string =
            json?.choices?.[0]?.delta?.content ??
            json?.choices?.[0]?.message?.content ??
            ''
          if (delta) {
            full += delta
            opts.onDelta?.(delta, full)
          }
        } catch (e) {
          if ((e as Error).message?.startsWith('ARK upstream:')) throw e
          // 容错：忽略坏的 chunk
        }
      }
    }
    opts.onDone?.(full)
    return full
  } catch (err) {
    if ((err as Error).name === 'AbortError') {
      opts.onDone?.(full)
      return full
    }
    opts.onError?.(err as Error)
    throw err
  } finally {
    reader.releaseLock()
  }
}
